packages/riverpod/lib/src/core/element.dart

```diff
@@ -1336,8 +1336,8 @@ mixin SyncProviderElement<ValueT> on ProviderElement<ValueT, ValueT> {
       case AsyncLoading(:final error?, :final stackTrace?) when value.retrying:
       case AsyncError(:final error, :final stackTrace):
         return $ResultError(error, stackTrace);
-      case AsyncLoading():
-        return null;
+      case AsyncLoading(:final value):
+        return value == null ? null : $ResultData(value);
```

Comment:
Instead add a separate case bellow the `AsyncData` one with:

```dart
case AsyncLoading(:final value, hasData: true):
```


Note:
Using `value == null` is unsafe, because the state could be volontarily `null`. 
Hence the `hasData` check instead.

You may have to do `value as T`